### PR TITLE
skipSpecies and time controlled chemistry activation (#9)

### PR DESCRIPTION
**Pull request by blttkgl**
**Description**
 
This PR introduces two main features to the loadBalancedChemistryModel to improve efficiency and control of chemistry computations:
 
1. Skip Species Functionality
 
- Adds a skipSpecies switch to dynamically deactivate species with negligible mass fractions.
- Introduces skipThreshold to control when a species is considered inactive.
- Automatically:
- Sets species inactive if their field is missing at initialization
- Deactivates species during runtime when max(Y) < skipThreshold
- Sets corresponding transport values to zero
- Ensures species are reactivated when chemistry is turned back on
- Synchronizes active/inactive species across processors
 
This reduces computational cost in cases with many low-impact species.
 
2. Time-Controlled Chemistry Activation
 
Adds support for:
 
- startTime
- endTime
- begin
- repeat
- Enables periodic or window-based activation of chemistry
- Handles both standard and wrap-around time intervals
- Provides a chemistry() switch that controls whether chemistry is active at a given timestep
 
3. Additional Improvements
 
- Ensures reaction rates (RR) are set to zero when chemistry is inactive
- Adds runtime logging indicating whether chemistry is active/inactive
- Removes leftover debug output
- Updates test scripts and tutorials to reflect new functionality
- Adds an example (SandiaD_LTS/Allrun_DLB) demonstrating:
 
a. skipSpecies usage
b. delayed chemistry activation
 
**Why this is useful**:
- Reduces chemistry cost in large mechanisms
- Provides fine control over when chemistry is solved
- Useful for ignition-delay setups, staged combustion, or startup transients
- Improves robustness when species fields are partially defined
 
**Notes**
- Default behavior is unchanged unless new entries are specified
- Backward compatible with existing cases

### CODE REVIEW

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H [hunk 1]
```diff
@@ -56,4 +56,51 @@ Foam::loadBalancedChemistryModel<ThermoType>::RR(const label i) {
...
+template <class ThermoType>
+inline Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::timeValue() const
+{
+    if (repeat_ > 0)
+    {
+        return begin_ + fmod(this->time().userTimeValue() - begin_, repeat_);
+    }
+    else
+    {
+        return this->time().userTimeValue();
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: "begin" only has effect when "repeat" is > 0

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H [hunk 2]
```diff
@@ -88,6 +96,18 @@ private:
         //- Tabulation method reference
         chemistryTabulationMethod& tabulation_;
 
+    	//- Optional start time for chemistry calculation
+	    const scalar startTime_;
+
+	    //- Optional end time for chemistry calculation
+	    const scalar endTime_;
+
+        //- Optional begin time for the chemistry calculation
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: Unclear what is the difference between begin and startTime

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H [hunk 3]
```diff
@@ -189,6 +209,12 @@ public:
 	//- Return non const access to the chemical source terms
         inline volScalarField::Internal& RR(const label i);
 
+    //- Calculate the time for chemistry activation/deactivation
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: I would reword a bit. Calculate local time coordinate for chemistry activation/deactivation?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H [hunk 4]
```diff
@@ -56,4 +56,51 @@ Foam::loadBalancedChemistryModel<ThermoType>::RR(const label i) {
...
+        if (currentTime >= startTime_ && currentTime <= endTime_)
+        {
+            return this->chemistry_;
+        }
+        else
+        {
+            return false;
+        }
+    }
+    else
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: Why do we need the case startTime > endTime? When do you use it?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 5]
```diff
@@ -49,6 +50,8 @@ Foam::loadBalancedChemistryModel<ThermoType>::
             this->mesh(),
             scalar(0.0)
         ),
+        resetSkipSpecies_(false),
+        skipThreshold_(this->lookupOrDefault("skipThreshold", 1e-5)),
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Should we make default smaller? We would like to only remove species mass fractions for which is effectively zero

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 6]
```diff
@@ -77,6 +91,28 @@ Foam::loadBalancedChemistryModel<ThermoType>::
...
+                    this->mesh().time().name(),
+                    this->mesh(),
+                    IOobject::NO_READ
+                );
+
+                // Check if the species file is provided, if not set inactive
+                // and NO_WRITE
+                if (!header.headerOk())
+                {
+                    this->thermo().setSpecieInactive(i);
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Won't the solver fail before we arrive to this check? If there is no specie file, then thermo mass fraction field will fail to create. @blttkgl have you tested it?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 7]
```diff
@@ -150,11 +186,76 @@ Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::solve
...
     {
-        return great;
+        for(label i = 0; i < this->nSpecie(); i++)
+        {
+            if(i == this->thermo().defaultSpecie())
+            {
+                continue;
+            }
+
+            const scalar maxY = gMax(this->Y()[i].oldTime());
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: We don't need to recalculate if a specie became active from inactive state, because it can't

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 8]
```diff
@@ -150,11 +186,76 @@ Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::solve
...
+
+            if(maxY < skipThreshold_)
+            {
+                this->thermo().setSpecieInactive(i);
+                const_cast<volScalarField&>(this->Y()[i]) == 0;
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: The error here will go to to defaultSpecie via [normaliseY()](https://github.com/OpenFOAM/OpenFOAM-dev/blob/master/applications/modules/multicomponentFluid/thermophysicalPredictor.C#L76). Do we even need setting it to 0? If we assume that we skip zero fields already, why make them "more 0"?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 9]
```diff
@@ -150,11 +186,76 @@ Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::solve
...
+            }
+            else
+            {
+                this->thermo().setSpecieActive(i);
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: How can a specie become active here? How can it increase its mass fraction when chemistry is off?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 10]
```diff
@@ -150,11 +186,76 @@ Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::solve
...
+                    this->thermo().setSpecieActive(i);
+                }
+            }
+
+        }
+        resetSkipSpecies_ = false;
     }
+    this->thermo().syncSpeciesActive();
+
+    if(!chemistry())
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:194]: @arintanen has ideas how to simplify this part

Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:194]: Do we need to do this every time step?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 11]
```diff
@@ -150,11 +186,76 @@ Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::solve
...
+            {
+                this->RR(j)[celli] = 0.0;
+            }
+        }
+        Info << "Chemistry is inactive" << endl;
+        return great;
+    }
+    else
+    {
+        Info << "Chemistry is active" << endl;
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: I would remove this print, but this is matter of taste

**Review by moreff — CHANGES_REQUESTED** [review]
I haven't tested the feature myself yet, but the general comments are below

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 10 above)
Comment (arintanen) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:194]: I think single loop over species and setting Y = zero is enough. In addition now the for loops are in wrong order causing jumping in memory

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 11 above)
Comment (arintanen) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Or maybe remove line 252 and change 257 to "Solving chemistry"?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 9 above)
Comment (arintanen) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Actually even a user defined fvsource cannot make it active since the specie equation is not solved and thus the field is never updated

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 8 above)
Comment (arintanen) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: If this line is removed then gMax can be changed to max, since later this->thermo().syncSpeciesActive(); is called that will sync the inactive species

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 7 above)
Comment (arintanen) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: This check for default specie could be removed and put everything inside condition if (thermo_.solveSpecie(i)) which would then only handle active species (-default)

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H (hunk 1 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: You are right, the logic should be changed to 

```
    if (repeat_ > 0)
    {
        return fmod(this->time().userTimeValue(), repeat_);
    }
    else
    {
        return this->time().userTimeValue();
    }
```

here begin_ refers to beginTime (https://github.com/OpenFOAM/OpenFOAM-dev/commit/b97f57c5805dfe244829fd65d0a0af66a8d18561), but more I think about it I realise it does not make sense to introduce it here, since it only helps with write and execute controls of function objects.

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H (hunk 2 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: See above, I think we can and should remove begin here, I think it served a purpose during testing but like you pointed out it has no use.

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H (hunk 3 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: //- Compute local chemistry activation time coordinate, I will make that change

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H (hunk 4 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: Thats a wrap around case.

Imagine you have an engine where you want to activate chemistry around injection event. You will have to define something like this:

```
startTime  700;
endTime  60;
repeat    720;

```
that condition takes care of this wrap around case. Does that make sense?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 5 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: >We would like to only remove species mass fractions for which is effectively zero

Not necessarily true. In spray combustion for instance unless you are interested in exhaust gas kinetics, you can remove a lot of species with small (but not non-zero) mass fractions in the system. 

In my tests and in engine simulations so far the sweet spot between performance and accuracy is around 1e-6. How about we drop it to 1e-6?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 7 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: In the original implementation the idea was that skipSpecie would be actively working for the entirety of a simulation, including the parts where the chemistry is on. 

But upon testing and reading up on similar implementations in different CFD solvers, I have come to the conclusion that during the period where chemistry active, we should solve and transport the entirety of the mechanism with all the species. That is why I changed the logic so that skipSpecies is only active, when chemistry is off. 

I think this line is a remainder from that original implementation, where skipSpecie would actually track accumulated mass fraction of combustion products during active chemistry period. Therefore it can be removed.

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 11 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Removed with c89123bcbf03095bf64a1f541f95e7cae008e9d8

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 5 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Yes, but this shouldn't be the default behaviour. This is a specific case for which you can define the treshold manually. In many problems it is dangerous to set an arbitrary treshold like this, which can cut-off e.g. radicals.
I would set it to `small`, so that by default only species with internalField set to 0 are disabled. Also, patch-values should also be checked in `max()` to avoid situation when in a simulation an inlet of a specie to empty domain fails

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H (hunk 4 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: We can move this logic to the constructor and change endTime to duration

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H (hunk 1 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: Solved in https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/9fae57c1b8a7ce50445abd68cca63797324928d6

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H (hunk 2 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: Solved in https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/9fae57c1b8a7ce50445abd68cca63797324928d6

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H (hunk 3 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: Solved in https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/61591400f46bbd6c8ff5239062a66b5cb7ff459f

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 5 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: skipThreshold set to small: https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/5d563b20577b129dc8409002373451f89fbc3e44

max is used instead of gMax, which also checks the boundaries: 85163bbfe3a35c5e79f4844efaeca0ff36ea64dd  @drhcelik  this should solve the issue Oulu was facing.

**Comment by blttkgl** [comment]
TODO: Add skipSpecie functionality definition to README

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 8 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: 85163bbfe3a35c5e79f4844efaeca0ff36ea64dd

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 9 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: 85163bbfe3a35c5e79f4844efaeca0ff36ea64dd

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 10 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:194]: @arintanen  can you check 85163bbfe3a35c5e79f4844efaeca0ff36ea64dd and confirm that this is now correct way to loop and set RR to zero?

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 7 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: 85163bbfe3a35c5e79f4844efaeca0ff36ea64dd

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H (hunk 4 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModelI.H]: fixed in https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/87377f7c9e663d9f9b39caab3a58c947c75e9e63

**Comment by arintanen** [comment]
I think the switch function for the chemistry must be reimplemented. Now if an user don't want to use skip species and defines chemistry off in dictionary, the switch function will return true, I think. It is also very complicated function and hard to follow.

Having these three variables (startime duration repeat) I propose replacing the whole function with something simple like this:
```
scalar t = timeValue() - startTime_*min(repeat, trunc(timeValue()/startTime_));
return this->chemistry_ && (t >= startTime_) && (t < startTime_ + duration);
```

**Comment by blttkgl** [comment]
> I think the switch function for the chemistry must be reimplemented. Now if an user don't want to use skip species and defines chemistry off in dictionary, the switch function will return true, I think. It is also very complicated function and hard to follow.
> 
> Having these three variables (startime duration repeat) I propose replacing the whole function with something simple like this:
> 
> ```
> scalar t = timeValue() - startTime_*min(repeat, trunc(timeValue()/startTime_));
> return this->chemistry_ && (t >= startTime_) && (t < startTime_ + duration);
> ```

I'll write a unittest that tests and ensures that the implementation is correct, you can then implement your proposed change. How does that sound?

**Comment by blttkgl** [comment]
@arintanen , [69b6720](https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/69b672042f4519495888bc93a4d41c2d0b7a023c) separates the chemistry activation into a helper class, similar to other helper classes, and implements a number of tests to ensure it works under different inputs. 

If you think there is a simpler way to handle the logic, feel free to have a go at it, and test it with the new unit tests.

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 6 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Check is gone

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 8 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: I'll open a new thread

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 10 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:194]: Looping and setting to Zero looks fine to me (though I doubt it is needed for Y)

File: src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H:1]: All the functionality from this file and redundant ChemistryActivation.C can be implemented in `Foam::loadBalancedChemistryModel<ThermoType>::chemistry()`. The only reason to have it separately is to enable unit tests, which doesn't justify this heavy overcomplication

File: src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+inline bool chemistryActive
+(
+    const scalar startTime,
+    const scalar duration,
+    const scalar repeat,
+    const scalar timeValue
+)
+{
+    if (startTime <= -great) return true;
+    if (duration <= scalar(0)) return false;
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H]: What is the point of these checks? I don't see the logic here. Validity of dictionary inputs should be checked in `read()`

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 6 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C]: Yes, we don't need to explicitly check the existence of a specie field. If the field does not exist, it should be explicitly set to inactive, and the active/inactive logic will handle species activation during runtime.

File: src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+inline scalar timeValueFromUserTime(const scalar userTime, const scalar repeat)
+{
+    if (repeat > 0)
+    {
+        return fmod(userTime, repeat);
+    }
+    return userTime;
+}
+
+inline bool chemistryActive
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H]: Currently the implementation is wrong and tests does not cover that. It incorrectly handles startTime.
For example, consider this test case (you can try it in unittests)
```
TEST_CASE("startTime/duration with repeat (non-wrapping) 1")
{
    const scalar start = 5.0;
    const scalar duration = 1;
    const scalar repeat = 3;

    CHECK(!chemistryActive(start, duration, repeat, 0.0));
    CHECK(!chemistryActive(start, duration, repeat, 4.0));
    CHECK(!chemistryActive(start, duration, repeat, 4.999));
    CHECK(chemistryActive(start, duration, repeat, 5.001));
    CHECK(chemistryActive(start, duration, repeat, 5.999));
    CHECK(!chemistryActive(start, duration, repeat, 6.001));
    CHECK(!chemistryActive(start, duration, repeat, 7.999));
    CHECK(chemistryActive(start, duration, repeat, 8.001));
    CHECK(chemistryActive(start, duration, repeat, 8.999));
    CHECK(chemistryActive(start, duration, repeat, 9.001));
}
```
The normalization in `timeValueFromUserTime` must take into account startTime.

However, I would get rid of all these functions in this file altogether. @arintanen proposed a two-line solution. On unit tests I'll comment separately

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C [hunk 14]
```diff
@@ -150,8 +157,47 @@ Foam::scalar Foam::loadBalancedChemistryModel<ThermoType>::solve
...
     scalar t_unbalance(0);
 
-    if(!this->chemistry_)
+    if(!chemistry() && skipSpecies_)
+    {
+        for(label i = 0; i < this->nSpecie(); i++)
+        {
+            if (this->thermo().solveSpecie(i))
+            {
+                const scalar maxY = max(this->Y()[i].oldTime()).value();
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:165]: @arintanen upon quick search I cannot find a proof that max() checks patches also. Do you have a reference? If no, I can search for it further

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H [hunk 15]
```diff
@@ -61,6 +62,9 @@ class loadBalancedChemistryModel
     public chemistryModel<ThermoType>
 {
 
+protected:
```
Comment (moreff) [inline src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: Is there any reason to have this specific switch protected?

File: tests/unittests/testChemistryActivation.C
Comment (moreff) [inline tests/unittests/testChemistryActivation.C:1]: 1. On the other comment I gave an example wheh these will fail. 
2. Checking exact boundary values in float is bad practice, as direct comparison of floats on equality is dangerous.
3. These tests are executed with MPI. Unit tests unrelated to load balancing should be separate.

As I mentioned in another comment, I would remove this test due to unnecessary complexity it is creating. Or simplify it by getting rid of a separate `ChemistryActivation` if you want to keep it.

**Review by moreff — CHANGES_REQUESTED** [review]

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C (hunk 14 above)
Comment (blttkgl) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.C:165]: Yes, it does. We tested it with Aleksi with the Sandia Flame D tutorial. If gMax is used it will complain that the boundary patch mass fractions do not add up to 1.0, but will max it works. You are welcome to test it further

**Comment by blttkgl** [comment]
Hey @arintanen @moreff,

Thanks for the comments.

https://github.com/Aalto-CFD/DLBFoam/pull/50#discussion_r3247252172 good point, the repeat handling should take `startTime` into account.

For the rest of the comments, I’ve provided the main implementation approach, so feel free to refactor and adjust the structure as needed while keeping the functionality unchanged, and merge to `main` once ready.

Bulut

**Comment by blttkgl** [comment]
https://github.com/Aalto-CFD/DLBFoam/pull/50/commits/ef3a1a96d4f658cbb9c28d00eb0ad4e6fda278b0

this should solve the issue you raised in https://github.com/Aalto-CFD/DLBFoam/pull/50#discussion_r3247252172

I see benefit in keeping unittests, but if it doesnt justify creating a ChemistryActivation class, feel free to eliminate it completely.

**Review by arintanen — COMMENTED** [review]
From my side we are ready to merge

File: src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H (hunk 12 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryActivation/ChemistryActivation.H]: removed

File: src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H (hunk 15 above)
Comment (moreff) [reply src/thermophysicalModels/chemistryModel/chemistryModel/loadBalancedChemistryModel/loadBalancedChemistryModel.H]: moved to private

**Review by moreff — APPROVED** [review]

**Comment by moreff** [comment]
@blttkgl I am merging the functionality, but as you noted above, there is no description in the README.md now
